1. After running the branch merge-fork-request via privnet seven node for two days, the reorg error still happens on seed node

INFO [02-28|18:55:31.652] Imported new potential chain segment     number=22392 hash=9e06f6..ee6ff6 blocks=1 txs=12 mgas=34.209 elapsed=2.591s      mgasps=13.203  snapdiffs=2.58MiB   triediffs=42.47MiB  triedirty=240.43MiB
**ERROR[02-28|18:55:31.661] Impossible reorg, please file an issue   oldnum=22392 oldhash=9e06f6..ee6ff6 oldblocks=0 newnum=22392 newhash=9e06f6..ee6ff6 newblocks=0**
INFO [02-28|18:55:31.675] Chain head was updated                   number=22392 hash=9e06f6..ee6ff6 root=82c28b..2f8f50 elapsed=14.437333ms



2. And node7 has the nil pointer exception issue:
INFO [02-28|21:38:28.530] Looking for peers                        peercount=7 tried=0 static=0
INFO [02-28|22:22:04.130] Imported new potential chain segment     number=23744 hash=dac3cf..0959c5 blocks=1 txs=0  mgas=0.000  elapsed=1h32m42.478s mgasps=0.000  age=1h32m47s  snapdiffs=2.12MiB   triediffs=41.65MiB  triedirty=0.00B
INFO [02-28|22:22:16.311] Looking for peers                        peercount=6 tried=2 static=0
INFO [02-28|22:22:45.346] Looking for peers                        peercount=6 tried=0 static=0
INFO [02-28|22:27:41.415] Looking for peers                        peercount=5 tried=0 static=0
INFO [02-28|22:30:09.895] Looking for peers                        peercount=5 tried=0 static=0
INFO [02-28|23:27:50.828] Looking for peers                        peercount=4 tried=0 static=0
INFO [02-28|23:38:55.935] Looking for peers                        peercount=1 tried=0 static=0
INFO [02-28|23:38:57.106] Block synchronisation started
INFO [02-28|23:38:57.147] Mining aborted due to sync
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x147610b]

goroutine 91530926 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fetchHead(0xc000214d20, 0xc00015a070)
        github.com/ethereum/go-ethereum/eth/downloader/downloader.go:831 +0x6b
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer(0xc000214d20, 0xc00015a070, {0xda, 0xc3, 0xcf, 0x9b, 0xbc, 0xfd, 0x29, 0xbe, ...}, ...)
        github.com/ethereum/go-ethereum/eth/downloader/downloader.go:558 +0x4a8
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).synchronise(0x0?, {0xc02723a0c0, 0x40}, {0xda, 0xc3, 0xcf, 0x9b, 0xbc, 0xfd, 0x29, ...}, ...)
        github.com/ethereum/go-ethereum/eth/downloader/downloader.go:523 +0x56f
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).LegacySync(0xc000214d20, {0xc02723a0c0, 0x40}, {0xda, 0xc3, 0xcf, 0x9b, 0xbc, 0xfd, 0x29, ...}, ...)
It could be the low memory issue of server, I'll try to test using 4 nodes again.

--- a/core/blockchain.go
+++ b/core/blockchain.go
@@ -2339,8 +2339,7 @@ func (bc *BlockChain) reorg(oldHead *types.Header, newHead *types.Header) error 
 		// the ancestor of new head while these two blocks are not consecutive
 		log.Info("Extend chain", "add", len(newChain), "number", newChain[0].Number, "hash", newChain[0].Hash())
 		blockReorgAddMeter.Mark(int64(len(newChain)))
-	} else {
-		// len(newChain) == 0 && len(oldChain) > 0
+	} else if len(newChain) == 0 && len(oldChain) > 0 {
 		// rewind the canonical chain to a lower point.
 		log.Error("Impossible reorg, please file an issue", "oldnum", oldHead.Number, "oldhash", oldHead.Hash(), "oldblocks", len(oldChain), "newnum", newHead.Number, "newhash", newHead.Hash(), "newblocks", len(newChain))
 	}

--- a/eth/downloader/downloader.go
+++ b/eth/downloader/downloader.go
@@ -828,7 +828,14 @@ func (d *Downloader) fetchHead(p *peerConnection) (head *types.Header, pivot *ty
 	mode := d.getMode()
 
 	// Request the advertised remote head block and wait for the response
+	p.lock.Lock()
+	if p.beacon == nil {
+		p.lock.Unlock()
+		return nil, nil, errUnknownPeer
+	}
 	latest, _ := p.beacon.Head()
+	p.lock.Unlock()
+
 	fetch := 1
 	if mode == SnapSync {
 		fetch = 2 // head + pivot headers

--- a/eth/downloader/peer.go
+++ b/eth/downloader/peer.go
@@ -83,10 +83,24 @@ func newPeerConnection(id string, version uint, peer ethPeer, logger log.Logger)
 
 // ConnectBeacon connects a downloader peer with a beacon peer.
 func (p *peerConnection) ConnectBeacon(peer beaconPeer) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
 	if p.beacon != nil {
 		return errAlreadyRegistered
 	} else {
 		p.beacon = peer
+		return nil
+	}
+}
+
+// DisconnectBeacon disconnects a downloader peer from its beacon peer.
+func (p *peerConnection) DisconnectBeacon() error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.beacon == nil {
+		return errNotRegistered
+	} else {
+		p.beacon = nil
 		return nil
 	}
 }
@@ -250,7 +264,10 @@ func (ps *peerSet) ConnectBeacon(id string, beacon beaconPeer) error {
 		ps.lock.Unlock()
 		return errNotRegistered
 	}
-	p.ConnectBeacon(beacon)
+	err := p.ConnectBeacon(beacon)
+	if err != nil {
+		return err
+	}
 	ps.lock.Unlock()
 
 	ps.events.Send(&peeringEvent{peer: p, join: true})
@@ -284,7 +301,10 @@ func (ps *peerSet) DisconnectBeacon(id string) error {
 		ps.lock.Unlock()
 		return errNotRegistered
 	}
-	p.beacon = nil
+	err := p.DisconnectBeacon()
+	if err != nil {
+		return err
+	}
 	ps.lock.Unlock()
 
 	ps.events.Send(&peeringEvent{peer: p, join: false})

